corehq/apps/userreports/specs.py

```diff
@@ -20,20 +19,70 @@ def TypeProperty(value):
     return StringProperty(required=True, choices=[value])
 
 
-@dataclass
-class FactoryContext:
-    named_expressions: dict
-    _named_expressions: dict = field(init=False, repr=False)
+class BaseContainer:
+    """Class used to manage access to named filters and expressions and
+    check for circular references.
+    """
+    def __init__(self, factory_context, expressions, domain=None):
+        self.factory_context = weakref.proxy(factory_context)
+        self.expressions = expressions
+        self.domain = domain
+        self.stack = []
+
+    def replace(self, expressions):
+        self.expressions = expressions
+
+    def get(self, name):
```

Comment:
This could be renamed to `__getitem__` to preserve the original dict item lookup inteface of `FactoryContext.named_expressions` and `FactoryContext.named_filters`.

Reply:
This is only called from within the `FactoryContext`. Users of the context call `get_named_filter` or `get_named_expression`. I specifically didn't make this mimic dict item lookup to make it easier to trace calls in code and to avoid assumptions that the underlying object is a dict.